mm2src/mm2_main/src/lp_wallet/mnemonics_wasm_db.rs

```diff
@@ -0,0 +1,163 @@
+use crypto::EncryptedMnemonicData;
+use mm2_core::mm_ctx::MmArc;
+use mm2_core::DbNamespaceId;
+use mm2_db::indexed_db::{DbIdentifier, DbInstance, DbUpgrader, IndexedDb, IndexedDbBuilder, OnUpgradeError,
+                         OnUpgradeResult, TableSignature};
+use mm2_err_handle::prelude::*;
+use std::collections::HashMap;
+
+type WalletsDBResult<T> = Result<T, MmError<WalletsDBError>>;
+
+#[derive(Debug, Deserialize, Display, Serialize)]
+pub enum WalletsDBError {
+    #[display(fmt = "Error deserializing '{}': {}", field, error)]
+    DeserializationError {
+        field: String,
+        error: String,
+    },
+    #[display(fmt = "Error serializing '{}': {}", field, error)]
+    SerializationError {
+        field: String,
+        error: String,
+    },
+    Internal(String),
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+struct MnemonicsTable {
+    wallet_name: String,
+    encrypted_mnemonic: String,
+}
+
+impl TableSignature for MnemonicsTable {
+    fn table_name() -> &'static str { "mnemonics" }
+
+    fn on_upgrade_needed(upgrader: &DbUpgrader, mut old_version: u32, new_version: u32) -> OnUpgradeResult<()> {
+        while old_version < new_version {
+            match old_version {
+                0 => {
+                    // do nothing explicitly because the table should be created on upgrade
+                    // from version 1 to 2 in order to avoid breaking existing databases
+                },
+                1 => {
+                    let table = upgrader.create_table(Self::table_name())?;
+                    table.create_index("wallet_name", true)?;
+                },
+                unsupported_version => {
+                    return MmError::err(OnUpgradeError::UnsupportedVersion {
+                        unsupported_version,
+                        old_version,
+                        new_version,
+                    })
+                },
+            }
+
+            old_version += 1;
+        }
+        Ok(())
+    }
+}
+
+pub(super) async fn save_encrypted_passphrase(
+    ctx: &MmArc,
+    wallet_name: &str,
+    encrypted_passphrase_data: &EncryptedMnemonicData,
+) -> WalletsDBResult<()> {
+    const DB_VERSION: u32 = 1;
+
+    // Create the database identifier
+    let db_name = "wallets";
+    let db_id = match ctx.db_namespace {
+        DbNamespaceId::Main => format!("MAIN::KOMODEFI::{}", db_name),
+        DbNamespaceId::Test(u) => format!("TEST_{}::KOMODEFI::{}", u, db_name),
+    };
+
+    let indexed_db_builder = IndexedDbBuilder {
```

Comment:
I can in all other cases IndexedDbBuilder is created with IndexedDbBuilder::new, maybe this is better?

Reply:
`new` function requires a `DbIdentifier` which includes the `wallet_rmd160`, so it's used for wallet level database. This DB is a multi-wallet db for storing mnemonics, we also can't know `wallet_rmd160` without getting the mnemonic first. Global wallets db implementation that doesn't depend on `wallet_rmd160` is on my checklist though https://github.com/KomodoPlatform/komodo-defi-framework/pull/2014#issue-2010065894 to have a specific builder for it and to keep it in the context, so it can be used in an RPC for the GUIs to retrieve the mnemonic when the user wants to view it.